uhdm-plugin/UhdmAst.cc

```diff
@@ -93,68 +93,62 @@ void UhdmAst::visit_range(vpiHandle obj_h,
-		// custom_var or int_var before any other AST_ASSIGN_EQ
-		// Especially before ones explicitly placed in initial block in source code
-		block_node->children.insert(block_child, 1, assign_node);
-		assign_node->children.push_back(id_node);
-		UhdmAst initial_ast(parent, shared, indent);
-		initial_ast.current_node = initial_node;
-		UhdmAst block_ast(&initial_ast, shared, indent);
-		block_ast.current_node = block_node;
-		block_ast.visit_one_to_one({vpiExpr},
-								   obj_h,
-								   [&](AST::AstNode* expr_node) {
-									   assign_node->children.push_back(expr_node);
-								   });
-	}
-}
-
-AST::AstNode* UhdmAst::process_value(vpiHandle obj_h) {
-	s_vpi_value val;
-	vpi_get_value(obj_h, &val);
-	std::string strValType;
-	if (val.format) { // Needed to handle parameter nodes without typespecs and constants
-		switch (val.format) {
-			case vpiScalarVal: return AST::AstNode::mkconst_int(val.value.scalar, false, 1);
+void UhdmAst::visit_default_expr(vpiHandle obj_h) {
+	UhdmAst initial_ast(parent, shared, indent);
+	UhdmAst block_ast(&initial_ast, shared, indent);
+    block_ast.visit_one_to_one({vpiExpr}, obj_h, [&](AST::AstNode *expr_node) {
+        auto mod = find_ancestor({AST::AST_MODULE});
+        AST::AstNode *initial_node = nullptr;
+        AST::AstNode *block_node = nullptr;
+        auto assign_node = new AST::AstNode(AST::AST_ASSIGN_EQ);
+        auto id_node = new AST::AstNode(AST::AST_IDENTIFIER);
+        id_node->str = current_node->str;
+
+        for (auto child : mod->children) {
+            if (child->type == AST::AST_INITIAL) {
+                initial_node = child;
+                break;
+            }
+        }
+        // Ensure single AST_INITIAL node is located in AST_MODULE
+        // before any AST_ALWAYS
+        if (initial_node == nullptr) {
+            initial_node = new AST::AstNode(AST::AST_INITIAL);
+            auto insert_it = find_if(mod->children.begin(), mod->children.end(), [](AST::AstNode *node) { return (node->type == AST::AST_ALWAYS); });
+            mod->children.insert(insert_it, 1, initial_node);
+        }
+        // Ensure single AST_BLOCK node in AST_INITIAL
+        if (initial_node->children.size() && initial_node->children[0]) {
+            block_node = initial_node->children[0];
+        } else {
+            block_node = new AST::AstNode(AST::AST_BLOCK);
+            initial_node->children.push_back(block_node);
+        }
+        auto block_child = block_node->children.begin();
+        for (; block_child != block_node->children.end(); block_child++) {
+            if ((*block_child)->type == AST::AST_ASSIGN_EQ) {
+                break;
+            }
+        }
+        // Insert AST_ASSIGN_EQ nodes that came from
+        // custom_var or int_var before any other AST_ASSIGN_EQ
+        // Especially before ones explicitly placed in initial block in source code
+        block_node->children.insert(block_child, 1, assign_node);
```

Comment:
Can we expect that the `block_child` was found?

Reply:
No, but then we just insert at the end.

Reply:
same as with the AST_ALWAYS considerations